File: bin/boj/BOJ_01780_TheNumberOfPaper.py
import sys


def divide_and_conquer(arr, row, col, size):
    global cnt_minus_1, cnt_0, cnt_1
    isSame = True
    first_color = arr[row][col]
    for i in range(row, row + size):
        for j in range(col, col + size):
            if paper[i][j] != first_color:
                isSame = False

    if isSame:
        if first_color == -1:
            cnt_minus_1 += 1

        elif first_color == 0:
            cnt_0 += 1

        else:
            cnt_1 += 1

    else:
        divided_size = size // 3
        divide_and_conquer(paper, row, col, divided_size)
        divide_and_conquer(paper, row, col + divided_size, divided_size)
        divide_and_conquer(paper, row, col + divided_size * 2, divided_size)
        divide_and_conquer(paper, row + divided_size, col, divided_size)
        divide_and_conquer(paper, row + divided_size,
                           col + divided_size, divided_size)
        divide_and_conquer(paper, row + divided_size, col +
                           divided_size * 2, divided_size)
        divide_and_conquer(paper, row + divided_size * 2, col, divided_size)
        divide_and_conquer(paper, row + divided_size * 2,
                           col + divided_size, divided_size)
        divide_and_conquer(paper, row + divided_size * 2,
                           col + divided_size * 2, divided_size)


N = int(sys.stdin.readline())
paper = []
cnt_minus_1 = 0
cnt_0 = 0
cnt_1 = 0

for _ in range(N):
    row = list(map(int, sys.stdin.readline().split()))
    paper.append(row)

divide_and_conquer(paper, 0, 0, N)

print(cnt_minus_1)
print(cnt_0)
print(cnt_1)

File: src/boj/BOJ_01780_TheNumberOfPaper.java
package boj;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class BOJ_01780_TheNumberOfPaper {
    static int cntMinus1 = 0;
    static int cnt0 = 0;
    static int cnt1 = 0;

    public static void divideAndConquer(int[][] arr, int row, int col, int size) {
        boolean isSame = true;
        int firstColor = arr[row][col];
        for (int i = row; i < row + size; i++) {
            for (int j = col; j < col + size; j++) {
                if (arr[i][j] != firstColor) {
                    isSame = false;
                    break;
                }
            }
            if (!isSame) {
                break;
            }
        }

        if (isSame) {
            if (firstColor == -1) {
                cntMinus1++;
            } else if (firstColor == 0) {
                cnt0++;
            } else {
                cnt1++;
            }
        } else {
            int dividedSize = size / 3;
            divideAndConquer(arr, row, col, dividedSize);
            divideAndConquer(arr, row, col + dividedSize, dividedSize);
            divideAndConquer(arr, row, col + dividedSize * 2, dividedSize);
            divideAndConquer(arr, row + dividedSize, col, dividedSize);
            divideAndConquer(arr, row + dividedSize, col + dividedSize, dividedSize);
            divideAndConquer(arr, row + dividedSize, col + dividedSize * 2, dividedSize);
            divideAndConquer(arr, row + dividedSize * 2, col, dividedSize);
            divideAndConquer(arr, row + dividedSize * 2, col + dividedSize, dividedSize);
            divideAndConquer(arr, row + dividedSize * 2, col + dividedSize * 2, dividedSize);
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());
        int[][] paper = new int[N][N];

        for (int i = 0; i < N; i++) {
            String[] row = br.readLine().split(" ");
            for (int j = 0; j < N; j++) {
                paper[i][j] = Integer.parseInt(row[j]);
            }
        }

        divideAndConquer(paper, 0, 0, N);

        System.out.println(cntMinus1);
        System.out.println(cnt0);
        System.out.println(cnt1);
    }
}
